File: ngff_zarr/multiscales.py
from dataclasses import dataclass
from typing import Any, Dict, List, Mapping, Optional, Sequence, Tuple, Union

from .methods import Methods
from .ngff_image import NgffImage
from .zarr_metadata import Metadata


@dataclass
class Multiscales:
    images: List[NgffImage]
    metadata: Metadata
    scale_factors: Optional[Sequence[Union[Dict[str, int], int]]] = None
    method: Optional[Methods] = None
    chunks: Optional[
        Union[
            int,
            Tuple[int, ...],
            Tuple[Tuple[int, ...], ...],
            Mapping[Any, Union[None, int, Tuple[int, ...]]],
        ]
    ] = None

File: ts/src/schemas/multiscales.ts
import { z } from "zod";
import { NgffImageOptionsSchema } from "./ngff_image.ts";
import { MetadataSchema } from "./zarr_metadata.ts";
import { MethodsSchema } from "./methods.ts";

export const ChunkSpecSchema: z.ZodTypeAny = z.union([
  z.number(),
  z.array(z.number()),
  z.array(z.array(z.number())),
  z.record(z.string(), z.union([z.number(), z.array(z.number())]).optional()),
]);

export const MultiscalesOptionsSchema: z.ZodTypeAny = z.object({
  images: z.array(NgffImageOptionsSchema),
  metadata: MetadataSchema,
  scaleFactors: z
    .array(z.union([z.record(z.string(), z.number()), z.number()]))
    .optional(),
  method: MethodsSchema.optional(),
  chunks: ChunkSpecSchema.optional(),
});
